File: level1/main.py
import os
import json

INPUT_FILE = os.path.dirname(__file__) + '/data/input.json'
OUTPUT_FILE = os.path.dirname(__file__) + '/data/output.json'


class User:
    def __init__(self, user, deals):
        self._id = user['id']
        self._name = user['name']
        # filter user deals
        self._deals = [deal for deal in deals
                       if deal['user'] == user['id']]

    def get_user_id(self):
        return self._id

    def get_commissions(self):
        commissions = 0
        # sum of all deals amount
        _sum = sum(deal['amount'] for deal in self._deals)
        # determine the commission in relation to the number of deals
        if len(self._deals) < 3:
            commissions = _sum * 0.10
        else:
            commissions = _sum * 0.20
        if _sum > 2000:
            commissions += 500
        return commissions


if __name__ == "__main__":
    with open(INPUT_FILE, 'r') as json_file:
        data = json.load(json_file)

    commissions = []
    for user in data['users']:
        _user = User(user, data['deals'])
        commissions.append({
            'user_id': _user.get_user_id(),
            'commission': _user.get_commissions(),
        })

    with open(OUTPUT_FILE, 'w') as output_file:
        json.dump({'commissions': commissions}, output_file)

    exit(0)

File: level1/main.js
'use strict'
const fs = require('fs')

const INPUT_FILE = './data/input.json'
const OUTPUT_FILE = './data/output.json'

function getCommissions(deals) {
  var commissions = 0
  // sum of all deals amount
  const sum = deals.reduce((a, b) => a + b.amount, 0)
  // determine the commission in relation to the number of deals
  commissions = deals.length < 3 ? (commissions = sum * 0.1) : (commissions = sum * 0.2)
  sum > 2000 && (commissions += 500)
  return commissions
}

function main() {
  const data = JSON.parse(fs.readFileSync(INPUT_FILE))
  var output = {}
  output.commissions = data.users.map((user) => ({
    user_id: user.id,
    // get commissions from the user deals
    commission: getCommissions(data.deals.filter((deal) => deal.user === user.id)),
  }))
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(output))
}

main()
